src/storage/mvcc/reader/forward_seeker.rs

```diff
@@ -0,0 +1,296 @@
+    lock_cursor: Cursor<S::Iter>,
+    write_cursor: Cursor<S::Iter>,
+
+    /// `default cursor` is lazy created only when it's needed.
+    default_cursor: Option<Cursor<S::Iter>>,
+
+    statistics: Statistics,
+}
+
+impl<S: Snapshot> ForwardSeeker<S> {
+    /// Take out and reset the statistics collected so far.
+    pub fn take_statistics(&mut self) -> Statistics {
+        ::std::mem::replace(&mut self.statistics, Statistics::default())
+    }
+
+    /// Get reference of the statics collected so far.
+    pub fn get_statistics(&self) -> &Statistics {
+        &self.statistics
+    }
+
+    /// Get the next key-value pair, where the key is greater or equal to given `key`.
+    pub fn read_next(&mut self, mut key: Key, ts: u64) -> Result<Option<(Key, Value)>> {
+        // TODO: Panic if given key is less than current key
+        // (Maybe we can panic in cursor's code)
+
+        let (mut write_valid, mut lock_valid) = (true, true);
+
+        // TODO: Add more comments to explain the logic.
+        loop {
+            let mut has_lock = false;
+            key = {
+                let (mut w_key, mut l_key) = (None, None);
+                // Try to advance both write_cursor and lock_valid
+                if write_valid {
+                    if self
+                        .write_cursor
+                        .near_seek(&key, &mut self.statistics.write)?
+                    {
+                        w_key = Some(self.write_cursor.key(&mut self.statistics.write));
+                    } else {
+                        w_key = None;
+                        write_valid = false;
+                    }
+                }
+                if lock_valid {
+                    if self.lock_cursor.near_seek(&key, &mut self.statistics.lock)? {
+                        l_key = Some(self.lock_cursor.key(&mut self.statistics.lock));
+                    } else {
+                        l_key = None;
+                        lock_valid = false;
+                    }
+                }
+                match (w_key, l_key) {
+                    (None, None) => return Ok(None),
+                    (None, Some(k)) => {
+                        has_lock = true;
+                        Key::from_encoded(k.to_vec())
+                    }
+                    (Some(k), None) => Key::from_encoded(k.to_vec()).truncate_ts()?,
+                    (Some(wk), Some(lk)) => if wk < lk {
+                        // Lock greater than `wk`, so `wk` must not have lock.
+                        Key::from_encoded(wk.to_vec()).truncate_ts()?
+                    } else {
+                        has_lock = true;
+                        Key::from_encoded(lk.to_vec())
+                    },
+                }
+            };
+            let lock = if has_lock {
+                Some(self.lock_cursor.value(&mut self.statistics.lock).to_vec())
+            } else {
+                None
+            };
+            if let Some(v) = self.get(&key, ts, lock)? {
```

Comment:
Maybe there is extra seek here in edge cases (also stands before the refine). For example, when `w_key` is None, we will seek a key which is `${user_key}_${ts}_${ts}`, and finally it doesn't exist. But we already know that.